clang/test/Driver/type-metadata.cpp

```diff
@@ -0,0 +1,145 @@
+// Test that type metadata are emitted with -fprofile-generate
+//
+// RUN: %clang -fprofile-generate -fno-lto -target x86_64-unknown-linux -emit-llvm -S %s -o - | FileCheck %s --check-prefix=ITANIUM
+// RUN: %clang -fprofile-generate -fno-lto -target x86_64-pc-windows-msvc -emit-llvm -S %s -o - | FileCheck %s --check-prefix=MS
```

Comment:
This test is forked from https://github.com/llvm/llvm-project/blob/main/clang/test/CodeGenCXX/type-metadata.cpp.

It's currently put under directory clang/test/Driver since `-fprofile-generate` is a driver option.

I'm wondering if it's more idiomatic to create the corresponding clang compiler options for `-fprofile-generate` for what I'm trying to do here? That way I could update `clang/test/CodeGenCXX/type-metadata.cpp` for test coverage.

Reply:
Yes I think that would be more typical and better imo (i.e. use %clang_cc to test)

Reply:
thanks! I updated the existing test (and removed the forked one) by only generating `!type` annotations iff clang cc1 option `-fprofile-instrument` is set to `llvm` (the FDO build).  CS-FDO and clang instrumentation sets this cc1 option differently, and it's intentional for this PR not to emit `!type` in these two settings mainly because there is no usage.

A longer story from some interesting findings when I tried to figured out how clang options work for the use case in this pr. Currently, clang driver accepts two options for FDO
   1.  `-fprofile-generate` (a [`Flag`](https://github.com/llvm/llvm-project/blob/c14e086b23929b0ac7092c45349a4f9749eda39c/clang/include/clang/Driver/Options.td#L1653) option that doesn't accept value according to https://clang.llvm.org/docs/InternalsManual.html#adding-new-command-line-option
  2. `-fprofile-generate=` (a [`Joined`](https://github.com/llvm/llvm-project/blob/c14e086b23929b0ac7092c45349a4f9749eda39c/clang/include/clang/Driver/Options.td#L1656) option that accepts the value 
  
I initially add [`CC1Option`](https://github.com/llvm/llvm-project/blob/c14e086b23929b0ac7092c45349a4f9749eda39c/clang/include/clang/Driver/Options.td#L64) to both (and added corresponding codegen options); and during this process found `-fprofile-instrument=llvm` is true if clang user either specified `-fprofile-generate` or `-fprofile-generate=path/to` (but not CS-FDO or clang instrumentation FDO),  so decided to just rely on this.